### ISSUE TITLE
Legacy hoisting for `function arguments() {}` still not correct?

### ISSUE THREAD

**Issue by anba** [tc39/ecma262#991]
refs: #815, #889, #891 

Per the current spec semantics:
```js
function f(a = () => arguments) {
  print(arguments);   // Print [object Arguments]
  print(a());         // Print [object Arguments]
  { function arguments() {} }
  print(arguments);   // Print function arguments() {}
  print(a());         // Print [object Arguments]
  
  delete arguments;
  print(arguments);   // Print [object Arguments]
  print(a());         // Print [object Arguments]
}
f();
```

Explanation:
1. `function arguments() {}` is an applicable legacy block-scoped function, because it can be replaced with `var arguments` without causing an early SyntaxError, and `arguments` is not a parameter name ([B.3.3.1 Changes to FunctionDeclarationInstantiation, step 1.a.ii](https://tc39.github.io/ecma262/#sec-web-compat-functiondeclarationinstantiation)).
1. [B.3.3.1 Changes to FunctionDeclarationInstantiation, step 1.a.ii.2](https://tc39.github.io/ecma262/#sec-web-compat-functiondeclarationinstantiation) does not create a binding in `varEnvRec`, because the function name is "arguments".
1. [9.2.12 FunctionDeclarationInstantiation, step 28](https://tc39.github.io/ecma262/#sec-functiondeclarationinstantiation) also doesn't create a binding for "arguments" in `varEnvRec`.
1. So when [B.3.3.1 Changes to FunctionDeclarationInstantiation, step 1.a.ii.3.f](https://tc39.github.io/ecma262/#sec-web-compat-functiondeclarationinstantiation) is executed, `fenvRec` doesn't have a binding for "arguments".
1. That means [8.1.1.1.5 SetMutableBinding, step 2.b](https://tc39.github.io/ecma262/#sec-declarative-environment-records-setmutablebinding-n-v-s) creates a deletable binding.

Maybe we just need something like this in B.3.3.1:
1. If instantiatedVarNames does not contain F, then
    1. Perform ! varEnvRec.CreateMutableBinding(F, false).
    1. If F is not "arguments" or argumentsObjectNeeded is false, let initialValue be undefined.
    1. Else,
        1. Let initialValue be ! envRec.GetBindingValue(F, false).
    1. Perform varEnvRec.InitializeBinding(F, initialValue).
    1. Append F to instantiatedVarNames.

And two nits in [B.3.3.1 Changes to FunctionDeclarationInstantiation](https://tc39.github.io/ecma262/#sec-web-compat-functiondeclarationinstantiation):
- Step 28 -> Step 29
- Typo: `initializedBindings`-> `instantiatedVarNames`.

**Comment by anba** [comment]
This test from test262 also asserts the wrong thing:
https://github.com/tc39/test262/blob/443b15e43e8c497b2edcf8c064b11fd7c1ad67fc/test/annexB/language/function-code/block-decl-func-skip-arguments.js#L34

**Comment by leobalter** [comment]
cc @rwaldron ^^

**Comment by littledan** [comment]
I'm not so happy with how complicated it turns out to support legacy hosting of functions named `argument`. Maybe we should just consistently skip this hoisting, as I think @syg was proposing at some point, and as those tests seem to assume.

**Comment by allenwb** [comment]
@littledan 
I'm with you. Legacy FiB supports solely to not break existing code that is interoperable among major browsers.  If there is evidence that not hoisting functions named arguments does not result in significant breakage then we should add it the complexity.  The rules for binding `arguments` is already too complicated.

**Comment by LeszekSwirski** [comment]
cc @syg
I was just going through some "delibarate incompatiblities" in V8, and came across this. Seems like the spec-incompatible hoisting is consistent across major implementations, is there any taste for changing this at some point?

**Comment by syg** [comment]
Yeah that strongly argues for codifying it in Annex B. Can't say I'd consider this to be high priority, but all other things being equal we should reflect reality.

**Comment by jmdyck** [comment]
> And two nits in [B.3.3.1 Changes to FunctionDeclarationInstantiation](https://tc39.github.io/ecma262/#sec-web-compat-functiondeclarationinstantiation):
> 
>     * Step 28 -> Step 29

Fixed by the merge of PR #1636 (and permanently fixed by #2052).

>     * Typo: `initializedBindings`-> `instantiatedVarNames`.

Fixed by the merge of PR #2951.
